source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -316,6 +304,118 @@ void RedisCluster::RedisDiscoverySession::startResolveRedis() {
 }
 
+void RedisCluster::RedisDiscoverySession::updateDnsStats(
+    Network::DnsResolver::ResolutionStatus status, bool empty_response) {
+  if (status == Network::DnsResolver::ResolutionStatus::Failure) {
+    parent_.info_->stats().update_failure_.inc();
+  } else if (empty_response) {
+    parent_.info_->stats().update_empty_.inc();
+  }
+}
+
+void RedisCluster::RedisDiscoverySession::resolveClusterHostnames(ClusterSlotsPtr&& slots) {
+  std::shared_ptr<ClusterSlotsPtr> s = std::make_shared<ClusterSlotsPtr>(std::move(slots));
+  resolveHostname(s, 0);
+}
+
+/**
+ * Resolve the primary cluster entry hostname in the slot. If the primary is already resolved,
+ * proceed to resolve replicas. Additionally, if the primary is successfully resolved, we proceed to
+ * resolve replicas.
+ *
+ * @param slots the list of slots which may need DNS resolution
+ * @param index the specific index into `slots` whose primary cluster entry needs to be resolved
+ */
+void RedisCluster::RedisDiscoverySession::resolveHostname(std::shared_ptr<ClusterSlotsPtr> slots,
+                                                          std::size_t index) {
+  if (index >= (**slots).size()) {
+    finishClusterHostnameResolution(std::move(slots));
+    return;
+  }
+  auto& slot = (**slots)[index];
+  if (slot.primary() == nullptr) {
+    ENVOY_LOG(trace,
+              "starting async DNS resolution for primary slot address {} at index location {}",
+              slot.primary_hostname_, index);
+    parent_.dns_resolver_->resolve(
+        slot.primary_hostname_, parent_.dns_lookup_family_,
+        [this, index,
+         slots = std::move(slots)](Network::DnsResolver::ResolutionStatus status,
+                                   std::list<Network::DnsResponse>&& response) -> void {
+          auto& slot = (**slots)[index];
+          ENVOY_LOG(trace, "async DNS resolution complete for {}", slot.primary_hostname_);
+          updateDnsStats(status, response.empty());
+          // If DNS resolution for a primary fails, we stop resolution for remaining, and reset the
+          // timer.
+          if (status != Network::DnsResolver::ResolutionStatus::Success) {
+            ENVOY_LOG(error, "Unable to resolve cluster slot primary hostname {}",
+                      slot.primary_hostname_);
+            resolve_timer_->enableTimer(parent_.cluster_refresh_rate_);
+            return;
+          }
+          // Primary slot address resolved
+          slot.setPrimary(Network::Utility::getAddressWithPort(
+              *response.front().addrInfo().address_, slot.primary_port_));
+          // Continue on to resolve replicas
+          resolveReplicas(std::move(slots), index);
+        });
+  } else {
+    resolveReplicas(std::move(slots), index);
+  }
+}
+
+/**
+ * Resolve the replicas in a cluster entry. If there are no replicas, simply move to the next
+ * cluster entry. Once the last replica is resolved, we move to resolving the next cluster entry.
```

Comment:
Per my other comment on concurrent resolution, why can't we resolve all the cluster entries at the same time if we always want to resolve all the addresses anyways?